flask/app/views/email.py

```diff
@@ -40,11 +40,30 @@ def subscribe():
                 return {'message': 'already'}, 200
             else:
                 lang = g.get('lang_code')
-                send_check_mail.delay(data_valide['email'], lang)
-                return {'message': 'send'}, 200
+                worker_task_response = send_check_mail.delay(data_valide['email'], lang)
+
+                if worker_task_response.status == "PENDING":
```

Comment:
I am not sure if checking "PENDING" is opposed or not when using flask. However I did not see such discouraging statement after googling a bit. Please feel free to let me know if there is a more idiomatic when using flask.

Reply:
To fix silent fail problem, maybe we can use retry function in celery?
```
    try:
        return send_mail(send_subject,
                     send_from,
                     send_to,
                     send_html_path,
                     check_email_path=check_email_path)
    except Exception as e:
        self.retry(countdown=5, exc=e, max_retries=2)
```
https://docs.celeryproject.org/en/stable/userguide/tasks.html#retrying

Reply:
After trying *retry*, *retry* does not dump the error message to flask log, which I originally expect to achieve. However, I sooner realized *result backend*, which is redis in our case when using TestConfig, has the corresponding celery task status and log etc. stored, and the information is sufficient for my original purpose: to make the website admin could trace system error in the future.

I will rollback this part of my code change back to the original code snippet.

Reply:
Re-pushed with the rollback and the corresponding success/failure logging test from the redis result backend.

```
$ docker exec -t -i redis_article /bin/sh
/data # redis-cli
127.0.0.1:6379> select 2
OK
127.0.0.1:6379[2]> keys *
1) "celery-task-meta-116b31cd-8889-4425-956c-c48b856c2916"
127.0.0.1:6379[2]> get "celery-task-meta-116b31cd-8889-4425-956c-c48b856c2916"
"{\"status\": \"SUCCESS\", \"result\": \"66\", \"traceback\": null, \"children\": [], \"date_done\": \"2020-11-26T22:37:48.566734\", \"task_id\": \"116b31cd-8889-4425-956c-c48b856c2916\"}"
127.0.0.1:6379[2]> 137
```

```
$ docker exec -t -i redis_article /bin/sh
/data # redis-cli
127.0.0.1:6379> select 2
OK
127.0.0.1:6379[2]> keys *
1) "celery-task-meta-e5300bfe-dda0-4955-8617-6fe18ed44b25"
127.0.0.1:6379[2]> get "celery-task-meta-e5300bfe-dda0-4955-8617-6fe18ed44b25"
"{\"status\": \"FAILURE\", \"result\": {\"exc_type\": \"SMTPResponseException\", \"exc_message\": [334, \"b'UGFzc3dvcmQ6'\"], \"exc_module\": \"smtplib\"}, \"traceback\": \"Traceback (most recent call last):\\n  File \\\"/usr/local/lib/python3.7/site-packages/celery/app/trace.py\\\", line 412, in trace_task\\n    R = retval = fun(*args, **kwargs)\\n  File \\\"/usr/local/lib/python3.7/site-packages/celery/app/trace.py\\\", line 704, in __protected_call__\\n    return self.run(*args, **kwargs)\\n  File \\\"/app/app/views/celery_tasks/tasks.py\\\", line 46, in send_check_mail\\n    check_email_path=check_email_path)\\n  File \\\"/app/app/views/celery_tasks/send_email.py\\\", line 60, in send_mail\\n    print(\\\"Error in sending emails: \\\", e)\\n  File \\\"/usr/local/lib/python3.7/smtplib.py\\\", line 280, in __exit__\\n    raise SMTPResponseException(code, message)\\nsmtplib.SMTPResponseException: (334, b'UGFzc3dvcmQ6')\\n\", \"children\": [], \"date_done\": \"2020-11-26T22:39:52.359458\", \"task_id\": \"e5300bfe-dda0-4955-8617-6fe18ed44b25\"}"
127.0.0.1:6379[2]>
```